---
title: "ad_analysis_code"
output: html_document
date: "2025-06-25"
---
#Libraries
```{r LOAD}
library(GenomicRanges)
library(tidyverse)
library(annotatr)
library(BSgenome.Hsapiens.UCSC.hg38)
library(Biostrings)
library(rtracklayer)
library(jsonlite)
library(ggplot2)
library(ggpubr)
library(biomaRt)
library(patchwork)
library(corrplot)
library(data.table)
library(BiocManager)
library(SNPlocs.Hsapiens.dbSNP155.GRCh38)
library(org.Hs.eg.db) 
library(AnnotationDbi)
```


#Generating AD SNP sequences for DeepCLip analysis
add the question bit at the end , but add janitor clean somewhere here 
```{r READ IN DATA}
ad_snps_start<- fread("ad_gwas.tsv", sep = "\t") 
```

```{r MAP SNPS TO ALLELES}
valid_snp_id <- grep("^rs\\d+$", ad_snps_start$SNPS, value = TRUE)
snp_info <- snpsById(SNPlocs.Hsapiens.dbSNP155.GRCh38, valid_snp_id, ifnotfound="drop") 
snp_alleles <- as.data.frame (snp_info)
snp_alleles_unique <- unique(snp_alleles, by = "RefSNP_id")
ad_gwas_annotated <- merge(ad_snps_start, snp_alleles_unique,
                           by.x = "SNPS", by.y = "RefSNP_id", all.x = TRUE)

view(ad_gwas_annotated)
iupac_codes <- c(
  "M" = "A/C", "K" = "G/T", "R" = "A/G", "Y" = "C/T",
  "S" = "G/C", "W" = "A/T", "B" = "C/G/T", "D" = "A/G/T"
)

ad_gwas_annotated$alleles <- iupac_codes[ad_gwas_annotated$alleles_as_ambig] #dont get values until ~line 94

```

```{r CREATE NEW SNP ONLY/ RISK ALLELE ONLY COLUMN}
ad_gwas_annotated_separated <- ad_gwas_annotated |> 
  separate(
    `STRONGEST SNP-RISK ALLELE`, 
    into = c("SNP_Name", "Risk_Allele"),
    sep = "-(?!.*-)",  # Negative lookahead: split on last hyphen
    convert = FALSE    # Optional: prevents automatic type conversion
  ) 
```

```{r REMOVE SOME VARIANTS}
ad_gwas_annotated_cleaned <- ad_gwas_annotated_separated |>
  mutate(CONTEXT = trimws(tolower(CONTEXT)))

removed_variants <- ad_gwas_annotated_separated |>
  filter(grepl("regulatory_region_variant|intergenic_variant|stop_gained|inframe_insertion", 
               CONTEXT))

ad_gwas_removed_context <- anti_join(
  ad_gwas_annotated_separated,
  removed_variants,
  by = "CONTEXT"
)
```

```{r GET STRAND INFO}
mapped_gene_symbols <- ad_gwas_removed_context$MAPPED_GENE # gets ENTREZ IDs

gene_ids <- mapIds(
  org.Hs.eg.db,
  keys = mapped_gene_symbols,
  keytype = "SYMBOL",
  column = "ENTREZID"
)

gene_id_character <- as.character(gene_ids)
gene_id_character <- gene_id_character[!is.na(gene_id_character)]

mapped_gene_strands <- AnnotationDbi::select(          #gets strand info from TxDb
  TxDb.Hsapiens.UCSC.hg38.knownGene,
  keys = gene_id_character,
  columns = "TXSTRAND",
  keytype = "GENEID"
)

strand_data <- mapped_gene_strands |>    #creates consensus strand per gene- highlights single dominant transcription direction - avoids confusion /misinterpretation]
  group_by(GENEID) |> 
  summarize(
    Strand = case_when(
      all(TXSTRAND == "+") ~ "+",
      all(TXSTRAND == "-") ~ "-",
      TRUE ~ "*"
    )
  ) |> #adds gene symbols back in 
  mutate(
    SYMBOL = mapIds(org.Hs.eg.db,
                    keys = GENEID,
                    keytype = "ENTREZID",
                    column = "SYMBOL")
  ) |> 
  dplyr::select(SYMBOL, Strand) |> 
  distinct(SYMBOL, .keep_all = TRUE) #removes duplicates 
            
     
strand_gwas_data <- ad_gwas_removed_context |> #merge 
  left_join(strand_data, by = c("MAPPED_GENE" = "SYMBOL"))

```

```{r MAKING OTHER ALLELE COLUMN}
filtering_alleles <- strand_gwas_data |>
  filter(!is.na(alleles)) |> 
  group_by(SNP_Name) |> 
  mutate(allele_count = length(unlist(strsplit(unique(alleles), split = "/")))) |> 
  filter(allele_count != 3) |> 
  ungroup() |> 
  filter(Risk_Allele != "?")

strand_gwas_data_separate_allele <- filtering_alleles |> 
  mutate( allele_list = strsplit(alleles, "/"),
          non_risk = map2_chr(                    # Extract the non-risk allele(s)
            allele_list,
            Risk_Allele,
            ~ paste(setdiff(.x, .y), collapse ="/") # Keep alleles that are NOT the risk allele
          )) |> 
  dplyr::select(-allele_list) |>             #removes temporary column
  mutate (is_risk_allele = ifelse(
    str_detect(alleles, fixed(Risk_Allele)),
    TRUE,
    FALSE
  )) |> 
  filter(is_risk_allele)
```

```{r JANITOR CLEAN NAMES}
strand_gwas_data_separate_allele <- strand_gwas_data_separate_allele |> 
  janitor::clean_names() 
```

```{r CREATE GRANGES}
strand_gwas_data_separate_allele <- strand_gwas_data_separate_allele |> 
  mutate(
    strand_2 = case_when(
      strand_2 %in% c("+", "-", "*") ~ strand_2,
      TRUE ~ "*"  # Set invalid values to unknown
    )
  )

strand_gwas_data_separate_allele$strand <- NULL

strand_gwas_data_separate_allele <- strand_gwas_data_separate_allele |> 
  rename(sequence_names = seqnames)  # needed to rename as GRange cant be created with seqnames (and strand) already in use 

ad_snps_gr <- strand_gwas_data_separate_allele |> 
    mutate(                                                        #check and clean CHR_POS
    chr_id = paste0('chr', chr_id),
    chr_pos = suppressWarnings(as.numeric(chr_pos)),                  # Convert to numeric and suppress the warning
    strand_2 = ifelse(strand_2 %in% c("+", "-"), strand_2, "*")     # Force valid strand values
  ) |> 
  filter(!is.na(chr_pos)) |># Remove problematic rows
  makeGRangesFromDataFrame(                  # Create GRanges object
    keep.extra.columns = TRUE,
    seqnames.field = "chr_id", 
    start.field = "chr_pos",
    end.field = "chr_pos",
    strand.field = "strand"
)

```

```{r ANNOTATING SNPS}
ad_normal_genes = build_annotations(genome = 'hg38', annotations = c("hg38_basicgenes")) 

ad_snp_annotated_gr <- annotate_regions(ad_snps_gr,annotations = ad_normal_genes) 
amigoingmad()
ad_snp_annotated <- as.data.frame(ad_snp_annotated_gr)



ad_snp_annotated_strand <- ad_snp_annotated |> 
 select(seqnames:snps, risk_allele, non_risk, annot.strand) |> 
  select(-strand) |> # selects but removes strand from it
  unique() |> # doesnt show repeated data / names 
  makeGRangesFromDataFrame(    keep.extra.columns = TRUE,
                               strand.field = 'annot.strand')



annotated_sequence <- getSeq(BSgenome.Hsapiens.UCSC.hg38, ad_snp_annotated_strand)
```

```{r SANITY CHECK}
ad_snp_annotated_strand |> 
  as.data.frame() |> 
  mutate(coding = ifelse(strand == "+",
                         non_risk,
                         as.character(reverseComplement((DNAStringSet(non_risk)))))) |> 
  mutate(extracted_sequence = as.character(annotated_sequence))  |> 
  filter(coding != extracted_sequence) |>  
  mutate( 
    reverse_complement_check = as.character (reverseComplement(DNAStringSet(non_risk))) == extracted_sequence)

table(ad_snp_annotated_strand$strand)
```

```{r RESIZE}
ad_snp_anotated_resize <- ad_snp_annotated_strand |> 
  resize (width = width(ad_snp_annotated_strand) + 74, fix = "center", ignore.strand = FALSE) # CHECK IF 74 CORRECT

```

```{r HEALTHY FLANK SEQUENCE FASTA}

ad_seq_flank = getSeq( BSgenome.Hsapiens.UCSC.hg38, ad_snp_anotated_resize)

ad_snp_annotated_strand$flank_sequence <- as.character(ad_seq_flank)

ad_snp_annotated_strand_df <- as.data.frame(ad_snp_annotated_strand)

colnames(ad_snp_annotated_strand_df)
ad_snp_annotated_strand_df |> 
  select(snps, flank_sequence) |> 
  view()

ad_healthy_DSS <- DNAStringSet(ad_snp_annotated_strand$flank_sequence)
names(ad_healthy_DSS) <- ad_snp_annotated_strand$snps

writeXStringSet(ad_healthy_DSS, filepath = "ad_healthy_seq_test.fasta")

```

```{r RISK FLANK SEQUENCE FASTA}
ad_flank_risk_seq <- ad_snp_annotated_strand_df |> 
  mutate(
    risk_coding_allele = ifelse(
      strand == "+",
      risk_allele,
      as.character(reverseComplement(DNAStringSet(risk_allele)))
    )
    ) |> 
  mutate(
    risk_flank = paste0(substr(ad_seq_flank, start = 1, stop = 37), risk_coding_allele,
                        substr(ad_seq_flank, start = 39, stop = 75))
  )

colnames(ad_flank_risk_seq)
ad_flank_risk_seq |> 
  select(snps, risk_flank) |> 
  view()
ad_flank_risk_seq$risk_flank

ad_flank_risk_DSS <- DNAStringSet(ad_flank_risk_seq$risk_flank)
names(ad_flank_risk_DSS) <- ad_flank_risk_seq$snps

writeXStringSet(ad_flank_risk_DSS, filepath = "ad_risk_test.fasta")
```

# AD DeepClip Binding Profiles Code 
```{r READ IN DATA}
ad_deepclip_data <- stream_in(file("C:/Users/Kai/Desktop/tdp_snp_analysis/data/ad_gwas1.json"))
```

```{r CLEAN THEME FOR BINDING PROFILES}
clean_theme <- function() {
  ggpubr::theme_pubclean() +     #creates clean theme to reuse in all plots
    theme(
      axis.line = element_line(colour = "black"),
      axis.text = element_text(colour = "black"),
      strip.text = element_text(face = "bold")
    )
}
```

```{r BINDING PROFILES}
ad_gwas1 <- jsonlite::fromJSON("C:/Users/Kai/Desktop/tdp_snp_analysis/data/ad_gwas1.json")$predictions

ad_gwas1$variant_sequence = gsub("T", "U", ad_gwas1$variant_sequence)  #makes sure its RNA sequence
ad_gwas1$variant_sequence = gsub("t", "u", ad_gwas1$variant_sequence)
ad_gwas1$sequence = gsub("T", "U", ad_gwas1$sequence)
ad_gwas1$sequence = gsub("t", "u", ad_gwas1$sequence)

# FUNCTION CREATION
paired_plot <- function(ad_gwas1, plot_difference) {      #    plot_difference =  function parameter (like a switch) that determines whether the plot shows: FALSE = raw weights for both sequences, TRUE = difference between weights
  weights1 <- unlist(x$weights)
  weights2 <- unlist(x$variant_weights)
  
  seq1 <- strsplit(toupper(ad_gwas1$sequence), "")[[1]]      # prepares code for comparison - toupper = all upper case, strsplit = splits character string into individual characters, [1] = select first element from each list returned by strsplit 
  seq2 <- strsplit(toupper(ad_gwas1$variant_sequence), "")[[1]]  # output = character vectors 
  
  if(plot_difference) {    #for plot_difference = TRUE
    weights2 <- weights2 - weights1  #calculates difference (variant - reference)
    tbl <- data.frame(
      pos = seq_along(seq2),
      weight = weights2,        #difference values
      group = factor(rep("difference", length(seq2))) # all rows labelled difference 
    )
  } else {      #for plot_difference = FALSE 
    tbl <- data.frame(
      pos = c(seq_along(seq1), seq_along(seq2)), # combines positions for both sequences 
      weight = c(weights1, weights2),     #stacks reference and variants weights 
      group = factor(c(rep("reference", length(seq1)), rep("variant", length(seq2))), levels=c("reference","variant"))
    )   #explicit factor levels 
  }
  
  
  xlabels <- mapply(function(a, b) paste(a, ifelse(a==b, "", b), sep="\n"), seq1, seq2) # visual comparison of sequences via stacking them vertically and highlighting differences 
  
  p <- ggplot(tbl, aes(pos, weight))
  if(plot_difference) p <- p + geom_hline(yintercept=0, color="dodgerblue") # only if plot_difference = TRUE - adds horizontal blue line at y intercept to show no difference - helps visualise positive and negative difference 
  p <- p +
    geom_line(aes(color=group), size=0.8) +
    scale_x_continuous(breaks=seq(1, max(tbl$pos)), labels=xlabels) +   #x-axis ticks at each position
    scale_color_manual(values=c("black", "red")) +
    clean_theme() +
    theme(
      legend.title = element_blank(),
      axis.title.x = element_blank(),
      axis.text.x = element_text(size=11)
    ) + labs(y="DeepCLIP score")
  return(p)
}

for (i in 1:dim(ad_gwas1)[1]) {  
  x = ad_gwas1[i,]   #extracts i-th row - assumes data in df where each row contains sequence data 
  width = 10.5   #sets pdf dimensions
  height = 3.65
  if (length(ad_gwas1$weights[[1]]) <= 30) {width = 7.75}
  p = paired_plot(ad_gwas1, plot_difference = FALSE)  #plot generation
  pdf(paste0("profile_",i,".pdf"), width = width, height = height)  #saves plot as pdf
  print(p)
  dev.off()
  
  p_diff = paired_plot(ad_gwas1, plot_difference = TRUE)
  pdf(paste0("profile_",i,".difference.pdf"), width = width, height = height)
  print(p_diff)
  dev.off()
}

```

# Finding Overlaps between DeepClip output and TDP-43 Binding Regions
```{r READ IN BED}
bed_data <- read.table(
  unz(here::here("data/postar3_tardbp_reduced.bed.zip"),
  "postar3_tardbp_reduced.bed"),
  sep = "\t",
  header = FALSE,
  stringsAsFactors = FALSE
)
```

```{r CONVERT DEEPCLIP INTO GRANGE}
ad_gwas1_df <- as.data.frame(ad_gwas1)


ad_snp_annotated_strand_df <- as.data.frame(ad_snp_annotated_strand) |> 
  rename(sequence_name = seqnames,
         start_ = start,
         end_ = end,
         width_ = width,
         strand_ = strand)

ad_gwas1_df <- ad_gwas1_df |> 
  rename(snps = id)|> 
  left_join(ad_snp_annotated_strand_df, by = "snps") |> #get chr and position info from ad_snp_annotated_strand
  dplyr::relocate(snps)  
  

ad_gwas_gr <- ad_gwas1_df |> 
  makeGRangesFromDataFrame(
    keep.extra.columns = TRUE,
    seqnames.field = "sequence_name",
    start.field = "start_",
    end.field = "end_",
    strand.field = "annot.strand"
  )
```


```{r CONVERT BED FILE TO GRANGE}
granges_bed <- GRanges(
  seqnames = bed_data$V1,
  ranges = IRanges(start = bed_data$V2 + 1, end = bed_data$V3),  # BED is 0-based, R is 1-based so need to allow for that 
  strand = ifelse(bed_data$V6 %in% c("+", "-"), bed_data$V6, "*")
)
```


```{r CHECKING STRAND INFO}
seqlevels(ad_gwas_gr) <- sub("^chr", "", seqlevels(ad_gwas_gr))

# Connect to Ensembl
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# Get genes overlapping  GRanges
genes <- getBM(
  attributes = c("chromosome_name", "start_position", "end_position", "strand"),
  filters = c("chromosomal_region"),
  values = paste0(seqnames(ad_gwas_gr), ":", start(ad_gwas_gr), "-", end(ad_gwas_gr)),
  mart = ensembl
)

# Convert numeric strands (1/-1) to character
genes$strand <- ifelse(genes$strand == 1, "+",
                       ifelse(genes$strand == -1, "-", "*"))

# Convert to GRanges and assign strands
ad_gwas_gr_strand <- makeGRangesFromDataFrame(
  genes,
  seqnames.field = "chromosome_name",  
  start.field = "start_position",     
  end.field = "end_position",        
  strand.field = "strand",            
  keep.extra.columns = TRUE           
)
hits <- findOverlaps(ad_gwas_gr, ad_gwas_gr_strand)
strand(ad_gwas_gr)[queryHits(hits)] <- strand(ad_gwas_gr_strand)[subjectHits(hits)]

# Convert chromosome names to UCSC format
uscs_format <- ifelse(
  seqlevels(ad_gwas_gr) == "MT",  # Handle mitochondrial case
  "chrM",
  paste0("chr", seqlevels(ad_gwas_gr))
)

# Apply new names
seqlevels(ad_gwas_gr) <- uscs_format 

```

some differences between strands in ad_snp_annotated and the new generated one, see if i need to remove the old one to avoid confusion/ missing certain snps 

```{r FINDING OVERLAPS}
ad_gwas_gr <- unique(ad_gwas_gr)

ad_binding_overlap <- subsetByOverlaps(ad_gwas_gr,
                                       granges_bed,
                                       maxgap = 200,
                                       ignore.strand = FALSE)
```

